File: client-backend/src/main/kotlin/controller/ArticleController.kt
package org.napbad.clientbackend.controller

import lombok.extern.slf4j.Slf4j
import org.babyfish.jimmer.Page
import org.babyfish.jimmer.client.meta.Api
import org.napbad.clientbackend.service.ArticleService
import org.napbad.model.dto.article.*
import org.napbad.utilities.log.logInfo
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@Api
@Slf4j
@CrossOrigin
@RestController
@RequestMapping("/article")
class ArticleController (
    val articleService: ArticleService
) {


    @Api
    @PostMapping("/add")
    fun add(
        @RequestBody article: ArticleCreateInput
    ): ArticleCreateOutput
    {
        if (article.title.isBlank()) {
            throw RuntimeException("title is null or blank")
        }
        logInfo("add article: ${article.title}")
        return articleService.add(article)
    }

    @Api
    @PostMapping("/query")
    fun getArticle(
        @RequestBody article: ArticleSpecification
    ): Page<ArticleQueryOutput>
    {
        return articleService.query(article)
    }

    @Api
    @PostMapping("/update")
    fun update(
        @RequestBody article: ArticleUpdateInput
    ): ArticleUpdateOutput
    {
        return articleService.update(article)
    }

    @Api
    @PostMapping("/delete")
    fun delete(
        @RequestBody article: ArticleDeleteInput
    ): String
    {
        return articleService.delete(article)
    }

}

File: client-frontend/src/api/services/ArticleController.ts
import type {Executor} from '../index.ts';
import type {
    ArticleCreateInput, 
    ArticleCreateOutput, 
    ArticleDeleteInput, 
    ArticleQueryOutput, 
    ArticleSpecification, 
    ArticleUpdateInput, 
    ArticleUpdateOutput, 
    Page
} from '../model/static';

export class ArticleController {
    
    constructor(private executor: Executor) {}
    
    readonly add: (options: ArticleControllerOptions['add']) => Promise<
        ArticleCreateOutput
    > = async(options) => {
        let _uri = '/article-account/add';
        return (await this.executor({uri: _uri, method: 'POST', body: options.body})) as Promise<ArticleCreateOutput>;
    }
    
    readonly delete: (options: ArticleControllerOptions['delete']) => Promise<
        string
    > = async(options) => {
        let _uri = '/article-account/delete';
        return (await this.executor({uri: _uri, method: 'POST', body: options.body})) as Promise<string>;
    }
    
    readonly getArticle: (options: ArticleControllerOptions['getArticle']) => Promise<
        Page<ArticleQueryOutput>
    > = async(options) => {
        let _uri = '/article-account/query';
        return (await this.executor({uri: _uri, method: 'POST', body: options.body})) as Promise<Page<ArticleQueryOutput>>;
    }
    
    readonly update: (options: ArticleControllerOptions['update']) => Promise<
        ArticleUpdateOutput
    > = async(options) => {
        let _uri = '/article-account/update';
        return (await this.executor({uri: _uri, method: 'POST', body: options.body})) as Promise<ArticleUpdateOutput>;
    }
}

export type ArticleControllerOptions = {
    'add': {
        readonly body: ArticleCreateInput
    }, 
    'getArticle': {
        readonly body: ArticleSpecification
    }, 
    'update': {
        readonly body: ArticleUpdateInput
    }, 
    'delete': {
        readonly body: ArticleDeleteInput
    }
}
